framework/ldtest/junit_test_logger.go

```diff
@@ -0,0 +1,206 @@
+	lock        sync.Mutex
+}
+
+type jUnitTestStatus struct {
+	failures  []error
+	skipped   ldvalue.OptionalString
+	output    string
+	startTime time.Time
+	duration  time.Duration
+}
+
+// Struct definitions for the JUnit XML schema - see https://github.com/jstemmer/go-junit-report
+
+type jUnitXMLDocument struct {
+	XMLName xml.Name            `xml:"testsuites"`
+	Suites  []jUnitXMLTestSuite `xml:"testsuite"`
+}
+
+type jUnitXMLTestSuite struct {
+	XMLName    xml.Name           `xml:"testsuite"`
+	Tests      int                `xml:"tests,attr"`
+	Failures   int                `xml:"failures,attr"`
+	Time       string             `xml:"time,attr"`
+	Name       string             `xml:"name,attr"`
+	Properties []jUnitXMLProperty `xml:"properties>property,omitempty"`
+	TestCases  []jUnitXMLTestCase `xml:"testcase"`
+}
+
+type jUnitXMLTestCase struct {
+	XMLName     xml.Name             `xml:"testcase"`
+	Classname   string               `xml:"classname,attr"`
+	Name        string               `xml:"name,attr"`
+	Time        string               `xml:"time,attr"`
+	SkipMessage *jUnitXMLSkipMessage `xml:"skipped,omitempty"`
+	Failure     *jUnitXMLFailure     `xml:"failure,omitempty"`
+}
+
+type jUnitXMLSkipMessage struct {
+	Message string `xml:"message,attr"`
+}
+
+type jUnitXMLProperty struct {
+	Name  string `xml:"name,attr"`
+	Value string `xml:"value,attr"`
+}
+
+type jUnitXMLFailure struct {
+	Message  string `xml:"message,attr"`
+	Type     string `xml:"type,attr"`
+	Contents string `xml:",chardata"`
+}
+
+func NewJUnitTestLogger(
+	filePath string,
+	serviceInfo harness.TestServiceInfo,
+	filters RegexFilters,
+) *JUnitTestLogger {
+	return &JUnitTestLogger{
+		filePath:    filePath,
+		serviceInfo: serviceInfo,
+		filters:     filters,
+		tests:       make(map[string]jUnitTestStatus),
+	}
+}
+
+func (j *JUnitTestLogger) TestStarted(id TestID) {
+	j.lock.Lock()
+	defer j.lock.Unlock()
+	j.testIDs = append(j.testIDs, id)
+	j.tests[id.String()] = jUnitTestStatus{
+		startTime: time.Now(),
+	}
+}
+
+func (j *JUnitTestLogger) TestError(id TestID, err error) {
+	j.lock.Lock()
+	defer j.lock.Unlock()
+	status := j.tests[id.String()]
+	status.failures = append(status.failures, err)
+	j.tests[id.String()] = status
+}
+
+func (j *JUnitTestLogger) TestFinished(id TestID, failed bool, debugOutput framework.CapturedOutput) {
+	j.lock.Lock()
+	defer j.lock.Unlock()
+	status := j.tests[id.String()]
+	status.output = debugOutput.ToString("")
+	status.duration = time.Since(status.startTime)
+	j.tests[id.String()] = status
+}
+
+func (j *JUnitTestLogger) TestSkipped(id TestID, reason string) {
+	j.lock.Lock()
+	defer j.lock.Unlock()
+	status := j.tests[id.String()]
+	status.skipped = ldvalue.NewOptionalString(reason)
+	j.tests[id.String()] = status
+}
+
+func (j *JUnitTestLogger) EndLog(results Results) error {
+	fmt.Printf("Writing JUnit data to %s\n", j.filePath)
+
+	var doc jUnitXMLDocument
+
+	properties := []jUnitXMLProperty{
```

Comment:
I think that in the context of actual JUnit, this "properties" list would represent Java system properties that are part of the test run configuration. I've made up some property names here just to capture some of the input parameters that affect the test harness .